lib/files.ts

```diff
@@ -59,10 +70,116 @@ export function checkForFileDownloadPath(path: string): boolean {
+  // loop.
+  derivedFromPaths = derivedFromPaths.filter(
+    (path) => !filePaths.includes(path) && !processedFileIds.has(path)
+  );
+
+  // Request these derived_from files from the server.
+  let derivedFromFiles: FileObject[] = [];
+  if (derivedFromPaths.length > 0) {
+    derivedFromFiles = (await requestFiles(
+      derivedFromPaths,
+      request
+    )) as FileObject[];
+  }
+
+  return derivedFromFiles;
+}
+
+/**
+ * Get all upstream files from the files in the current file set, based on their `derived_from`
+ * paths. We have to iteratively request the `derived_from` files for each batch of files until we
+ * have reached the top of the chains. `derived_from` doesn't contain embedded file objects, so we
+ * have to ascend these change batch by batch, collecting each batch's `derived_from` arrays.
+ *
+ * Depending on where they occur, either filter out or detect loops in the `derived_from` chains,
+ * which should tell the user to fix the data.
+ * @param files Array of files in a file set to get all upstream files for
+ * @param request Request object
+ * @returns Array of all upstream files from the files in the current file set
+ */
+export async function getAllDerivedFromFiles(
+  files: FileObject[],
+  request: FetchRequest
+) {
+  // Initialize `derivedFromFiles` collects all the upstream files from the files in the current
+  // file set. `processedFileIds` is a set of file IDs that have already been processed in the graph
+  // to prevent infinite loops in the case of loops in the `derived_from` chains.
+  let derivedFromFiles: FileObject[] = [];
+  const processedFileIds = new Set<string>();
+
+  // Prime the pump by getting the `derived_from` files for the files directly in the current file
+  // set.
+  let derivedFromFileBatch = await getFileDerivedFromFiles(
+    files,
+    request,
+    processedFileIds
+  );
+
+  // As long as we find derived-from files in the batch of files, keep going up the chain of these
+  // files until we have run out of `derived_from` files in a batch.
+  while (derivedFromFileBatch.length > 0) {
+    derivedFromFiles = derivedFromFiles.concat(derivedFromFileBatch);
+
+    // Add the current batch of file IDs to the processed set. This helps us detect loops in the
+    // `derived_from` chains. Throw an error if we detect a loop here.
+    derivedFromFileBatch.forEach((file) => {
+      if (processedFileIds.has(file["@id"])) {
+        throw new Error(
+          `Detected a derived_from loop with file ID: ${file["@id"]}`
+        );
+      }
+      processedFileIds.add(file["@id"]);
+    });
+
+    // We now have a batch of files that might derive from other files. Request any derived from
+    // files in this batch, which gets us a new upstream batch of files to process.
+    derivedFromFileBatch = await getFileDerivedFromFiles(
+      derivedFromFileBatch,
+      request,
+      processedFileIds
+    );
```

Comment:
is there a situation where the recursion level here can go too deep? not sure if it makes sense to add a limit to the recursion level

Reply:
I can’t _think_ of a situation where it could go too deep. It has detected loops I’ve set up, including multi-node loops. And I’m not sure how deep we can practically expect these to go right now, so if we chose a depth, I have no recommendation what that depth should be.

Reply:
sounds good, we can adapt in the future if we see any complicated cases to draw on the screen